utilities/write_batch_with_index/write_batch_with_index_internal.cc

```diff
@@ -155,34 +156,105 @@ int WriteBatchEntryComparator::CompareKey(uint32_t column_family,
   }
 }
 
+WriteEntry WBWIIteratorImpl::Entry() const {
+  WriteEntry ret;
+  Slice blob, xid;
+  const WriteBatchIndexEntry* iter_entry = skip_list_iter_.key();
+  // this is guaranteed with Valid()
+  assert(iter_entry != nullptr &&
+         iter_entry->column_family == column_family_id_);
+  auto s = write_batch_->GetEntryFromDataOffset(
+      iter_entry->offset, &ret.type, &ret.key, &ret.value, &blob, &xid);
+  assert(s.ok());
+  assert(ret.type == kPutRecord || ret.type == kDeleteRecord ||
+         ret.type == kSingleDeleteRecord || ret.type == kDeleteRangeRecord ||
+         ret.type == kMergeRecord);
+  return ret;
+}
+
+bool WBWIIteratorImpl::MatchesKey(uint32_t cf_id, const Slice& key) {
+  if (Valid()) {
+    return comparator_->CompareKey(cf_id, key, Entry().key) == 0;
+  } else {
+    return false;
+  }
+}
+
+WriteBatchWithIndexInternal::WriteBatchWithIndexInternal(
+    DB* db, ColumnFamilyHandle* column_family)
+    : db_(db), db_options_(nullptr), column_family_(column_family) {
+  if (db_ != nullptr && column_family_ == nullptr) {
+    column_family_ = db_->DefaultColumnFamily();
+  }
+}
+
+WriteBatchWithIndexInternal::WriteBatchWithIndexInternal(
+    const DBOptions* db_options, ColumnFamilyHandle* column_family)
+    : db_(nullptr), db_options_(db_options), column_family_(column_family) {}
+
+WriteBatchWithIndexInternal::WriteBatchWithIndexInternal(
+    ColumnFamilyHandle* column_family)
+    : db_(nullptr), db_options_(nullptr), column_family_(column_family) {}
```

Comment:
What if column_family is nullptr as in default? Below I see `Status::InvalidArgument("Must provide a column_family")`

Reply:
If there is no column family supplied, you cannot do merge operations because there the merge operator is associated with the column family.  This is the documented behavior in include/rocksdb/utilitiies/write_batch_with_index.h:
 // Similar to previous function but does not require a column_family.
  // Note:  An InvalidArgument status will be returned if there are any Merge
  // operators for this key.  Use previous method instead.
  Status GetFromBatch(const DBOptions& options, const Slice& key,
                      std::string* value) {
    return GetFromBatch(nullptr, options, key, value);
  }

Reply:
Why does this constructor fill in an empty column family and the others leave it nullptr? If the constructors are materially different, there need to be comments clarifying that. These are new constructors.

Reply:
There are three constructors,
- DB + ColumnFamilyHandle
- DBOptions + ColumnFamilyHandle
- ColumnFamilyHandle
For any of the constructors, ColumnFamilyHandle could theoretically be nullptr. When the argument is not specified (e.g. DB for the DBOptions constructor), the unspecified argument could be nullptr.

This could all be done via one constructor taking all three arguments and specifying nullptr for the unused argument but I do not know that helps with code clarity at all.  If you feel that it does, I can make that change.  There is no material difference between the constructors other than not needed to specify nullptr and remember the order of the arguments.